include/vkwrap/vkwrap/mman.h

```diff
@@ -0,0 +1,556 @@
+#pragma once
+
+#include "common/vulkan_include.h"
+#include <vk_mem_alloc.h>
+
+#include "utils/misc.h"
+#include "vkwrap/core.h"
+#include "vkwrap/error.h"
+
+#include <range/v3/algorithm/any_of.hpp>
+
+#include <range/v3/range/conversion.hpp>
+
+#include <array>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+namespace std
+{
+
+template <> struct hash<vk::Buffer>
+{
+    size_t operator()( const vk::Buffer& buffer ) const
+    {
+        return reinterpret_cast<size_t>( buffer.operator VkBuffer() );
+    } // operator()
+
+}; // struct hash<vk::Buffer>
+
+template <> struct hash<vk::Image>
+{
+    size_t operator()( const vk::Image& image ) const
+    {
+        return reinterpret_cast<size_t>( image.operator VkImage() );
+    } // operator()
+
+}; // struct hash<vk::Image>
+
+} // namespace std
+
+namespace vkwrap
+{
+
+inline bool
+hasStencil( vk::Format format )
+{
+    using vk::Format;
+
+    constexpr auto k_formats_with_stencil = std::array{
+        Format::eS8Uint,
+        Format::eD16UnormS8Uint,
+        Format::eD24UnormS8Uint,
+        Format::eD32SfloatS8Uint,
+    };
+
+    auto isEqualFormats = [ format ]( Format format_with_stencil ) {
+        return format == format_with_stencil;
+    };
+
+    return ranges::any_of( k_formats_with_stencil, isEqualFormats );
+} // hasStencil
+
+inline vk::ImageAspectFlags
+chooseAspectMask( vk::Format format, vk::ImageLayout new_layout )
+{
+    using vk::ImageAspectFlagBits;
+    using vk::ImageAspectFlags;
+    using vk::ImageLayout;
+
+    if ( new_layout != ImageLayout::eDepthStencilAttachmentOptimal )
+    {
+        return ImageAspectFlagBits::eColor;
+    }
+
+    ImageAspectFlags aspect_mask{ ImageAspectFlagBits::eDepth };
+
+    if ( hasStencil( format ) )
+    {
+        aspect_mask |= ImageAspectFlagBits::eStencil;
+    }
+
+    return aspect_mask;
+} // chooseAspectMask
+
+inline std::pair<vk::AccessFlags, vk::AccessFlags>
+chooseAccessMasks( vk::ImageLayout old_layout, vk::ImageLayout new_layout )
```

Comment:
It is unclear what the return type is. Prefer a POD struct with members (that have clear names) to ```std::pair```. This way it is obvious what first and second ```vk::AccessFlags``` means

Reply:
Something like:
```cpp
struct AccessMasksResult
{
  vk::AccessFlags src;
  vk::AccessFlags dst;
};
```

Reply:
Partly agree. In some cases (like this) it is not so important to add struct. The function is called in one place and caller's variables describe what result is. For greater clarity I've refactored it.